Refactoring attempt two

Katerina-codes opened the pull request:
My work on video kata from yesterday afternoon. Any general feedback or feedback on ways to move forward that are different to Robert Martin's approach would be greatly appreciated.

@MollieS @Maikon @jsuchy @felipesere

Katerina-codes committed:
Added /out to .gitignore

Katerina-codes committed:
Corrected indentation

Katerina-codes committed:
Break string at each new line for clear formatting

Katerina-codes committed:
Rename variable to amountOwed

- Moved variable to the top of the class.

Katerina-codes committed:
Extract important values to fields

- Add new getter methods to get these values.

Katerina-codes committed:
Apply refactoring to all tests

- Inline values to remove duplication.

Katerina-codes committed:
Extract switch statement to method

- Extract to amountAccruedForRentals method.
- Remove dead code.

Katerina-codes committed:
Rename variable each to eachMovie

Katerina-codes committed:
Remove testcase extension. Use Junit test setup

Katerina-codes committed:
Extract field from string and movie type 'The Cell'

Katerina-codes committed:
Rename field so it is clearer

Katerina-codes committed:
Extract all string and movie types to fields

Katerina-codes committed:
Add field variables to Before setup

Katerina-codes committed:
Replace vector with array

- Replace while loop with for each loop.

Katerina-codes committed:
Extract frequent renter points logic to method

Katerina-codes committed:
Move public variables above private method

Katerina-codes committed:
Extract logic to get thisAmount to own method

- Re-structure code by grouping logic.

Katerina-codes committed:
thisAmount renamed currentAmount for clarity

- Remove dead code.
- Restructure code to move private methods down to the bottom of the class.

Maikon commented on .gitignore:
```diff
@@ -17,3 +17,4 @@ bld/
 [Bb]in/
 [Oo]bj/
 _Resharper.videostore/
+out/
```
Missing new line.

Maikon commented on java/videostore/src/VideoStoreTest.java:
```diff
@@ -1,38 +1,61 @@
+        assertEquals(7.5, customer.getAmountOwed());
+        assertEquals(3, customer.getFrequentRenterPoints());
+    }
+}
```
Missing new line.

Maikon replied:
Good job on cleaning up the tests. I think you can go a step further by extracting helper functions for creating a new movie and rental. The `new` keyword appears a bit too much in this file.

Maikon commented on java/videostore/src/Customer.java:
```diff
@@ -1,67 +1,81 @@
+
+    private double getAmountAccrued(Rental eachMovie, double currentAmount) {
+        return amountAccruedForRentals(currentAmount, eachMovie);
+    }
```
This method doesn't add much since it's delegating directly to `amountAccruedForRentals`. Why not call `amountAccruedForRentals` directly instead?

felipesere commented on java/videostore/src/Customer.java:
```diff
@@ -1,67 +1,81 @@
+    }
+
+    private double getAmountAccrued(Rental eachMovie, double currentAmount) {
+        return amountAccruedForRentals(currentAmount, eachMovie);
```
This is interesting!
The only thing that `getAmountAccrued(...)` does is flip the arguments for `amountAccruedForRentals(...)`.
Doe you see any other value this method brings?
Otherwise I'd suggestion lining up the arguments and removing `getAmountAccrued(...)`

Katerina-codes replied:
Oh dear, well spotted. Thank you, this now fixed.

Maikon commented on java/videostore/src/Customer.java:
```diff
@@ -1,67 +1,81 @@
+    private double getAmountAccrued(Rental eachMovie, double currentAmount) {
+        return amountAccruedForRentals(currentAmount, eachMovie);
+    }
+    private void getFrequentRenterPoints(Rental eachMovie) {
```
The `get` prefix here is misleading because this method does not return anything. It has a _side-effect_ instead. This means it changes the state of the program in a non-obvious way. See the calls to `frequentRenterPoints++` below.

Maikon replied:
I think `movie` would be a better variable name than `eachMovie`.

felipesere commented on java/videostore/src/Customer.java:
```diff
@@ -1,67 +1,81 @@
+    private double getAmountAccrued(Rental eachMovie, double currentAmount) {
+        return amountAccruedForRentals(currentAmount, eachMovie);
+    }
+    private void getFrequentRenterPoints(Rental eachMovie) {
```
This method is a little _deceptive_.
It says "getFrequentRenterPoints" which lead me to believe that I will _get the frequent renter points_ for the movie I put in. In *reality*, it returns nothing and updates a global variable.
I would strongly suggest return the necessary frequent renter points and do the addition to the global "frequentRenterPoints" wherever you call this method.

Katerina-codes replied:
I have changed this now. I'm not 100% sure what you meant with regards to the addition as I thought it *was* being done on the global variable but the method now returns the frequentRenterPoints.

felipesere commented on java/videostore/src/Customer.java:
```diff
@@ -1,67 +1,81 @@
+            frequentRenterPoints++;
+    }
+
+    private double amountAccruedForRentals(double currentAmount, Rental each) {
```
Funny: this method returns a value, but does not have the `get` prefix 😸 .
This one actually deserves the `get` prefix though!
As with frequentRenterPoints, I'd suggest separating "calculating how many points this Rental is worth" form "updating the global variable _currentAmount_".

Maikon replied:
`each` is not very descriptive/informative. The effect of naming the rental variable `each` is more obvious when you look at this: `each.getMovie()`
- What does `each` mean?
- What kind of `each` has a movie attached to it?

The above questions are easily answered when you name the variable to what it is `rental`. The problematic line then changes to `rental.getMovie()` which tells me all that I need to know.

Maikon reviewed (commented):
Overall good first steps 👍

felipesere commented:
Hi @Katerina-codes 
Good initial refactoring. I see you isolated some chunks out that might want a new home 😄 
I have a feeling that there are a couple of refactoring that you can do now that the code is a bit more structured. For example, the frequent renter points and price can be sliced in a way, that they _only_ depend on the Rental/Movie. That _could_ allow us move that logic out of the statement into the Movies themselves.
You could also see if you can use inheritence/polymophism to replace the different checks on `priceCode`. The big benefit is that you can open the `KidsMovie` file and see how many points you get an what it costs, without having to look around multiple things...

Let me know if you need more info on anything.

Katerina-codes committed:
Extract money owed & renter points logic to method

Katerina-codes committed:
Remove unnecessary method

Katerina-codes committed:
getFrequentRenterPoints returns points

- Made if statement more explicit

Katerina-codes committed:
Rename method so it states that it returns points

Katerina-codes committed:
Removed global variable from method

felipesere commented on java/videostore/src/Customer.java:
```diff
@@ -1,67 +1,81 @@
+
+    private int getFrequentRenterPoints(Rental eachMovie) {
+        if (eachMovie.getMovie().getPriceCode() == Movie.NEW_RELEASE && eachMovie.getDaysRented() > 1) {
+            frequentRenterPoints += 2;
```
This is still a dependency on a global variable

felipesere commented on java/videostore/src/Customer.java:
```diff
@@ -1,67 +1,81 @@
+    private double getAmountAccruedForRentals(double currentAmount, Rental each) {
+        switch (each.getMovie().getPriceCode()) {
+            case Movie.REGULAR:
+                currentAmount += 2;
```
This is also a dependency on a global/external variable.
This method should only return the points that this single Rental awards.

Katerina-codes committed:
Remove currentAmount parameter

Katerina-codes committed:
Replace currentAmount global variable

- Replaced with points local variable.

Katerina-codes committed:
Extract get points for regular movie to method

Katerina-codes committed:
Rename argument each to rental for clarity

Katerina-codes committed:
Extract get points for New Release movie to method

Katerina-codes committed:
Extract get points for Childrens movie to method

Katerina-codes committed:
Remove extra line

Katerina-codes committed:
Remove instance variables from methods

Katerina-codes committed:
Rename methods for better clarity

Katerina-codes committed:
Remove amountOwed & FrequentRenterPoints variables

- Remove mutable objects from Customer class by converting instance variables to local variables and moving them to their own methods.
- Rename tests.
- Introduce test that tests whole statement.
- Rename methods and variables for better clarity.

Katerina-codes committed:
Move rental types & getAmountForRental to Rental

Katerina-codes committed:
Moved each rental type to its own class

- Removed each rental method's rental parameter and passed in days
rented.

Katerina-codes committed:
Remove dead code

Katerina-codes committed:
Each movie type extends class Movie

Katerina-codes committed:
Reformat tests

Katerina-codes committed:
Create new getAmountForRental without parameter

Katerina-codes committed:
Remove all usages of previous method call

Katerina-codes committed:
Inline old method functionality to new method

Katerina-codes committed:
Remove static from getAmount method

- Makes Movie class abstract.
- Makes getAmount method on Movie abstract.

Katerina-codes committed:
Movie calculates its own frequent renter points

- Moves getFrequentRenterPoints to Rental class.
- Extract calculation logic for each movie to new getFrequentRenter
points method on each move type class.

Katerina-codes committed:
Remove duplicated method by pulling method up

Katerina-codes committed:
Correct indentation

Katerina-codes committed:
Correct indentation
